File: src/07-B-criminality.R
library(arrow)
library(tidyverse)
library(readxl)

ARR2COM <- read_feather("results_building/arr2com.feather") 

DICO_CRIMES <- data.frame(
  var_crimes = c("coups", "trafic", "vols_av", "vols_sv", "cambriolages"), 
  crimes_desc = c("Coups et blessures volontaires", "Trafic de stupéfiants", "Vols violents sans arme", "Vols sans violence contre des personnes", "Cambriolages de logement")
)

TABLE_PASSAGE_DF <- read_feather("results_building/t_passage.feather") |>
  select(code_insee24, arr24, bv2022) |>
  unique() |>
  group_by(code_insee24) |>
  filter(row_number()==1) |>
  ungroup()

CRIMES_DF <- read_parquet("data/external/miom-crimes/crimes23_geo24.parquet") |>
  filter(annee>16)

CRIMES_DF <- CRIMES_DF |>
  select(code_insee24 = CODGEO_2024, 
         year = annee, 
         crime_count = faits,
         crimes_desc = classe,
         units = unité.de.compte) |>
  mutate(
    year = year+2000, 
    crime_count = if_else(is.na(crime_count), 0, crime_count)
  ) |>
  filter(crimes_desc%in%DICO_CRIMES$crimes_desc) 

CRIMES_DF <- CRIMES_DF |>
  left_join(ARR2COM, by = c("code_insee24" = "code_arr")) |>
  mutate(code_insee24 = if_else(is.na(code_com), code_insee24, code_com)) |>
  left_join(DICO_CRIMES, by = "crimes_desc") |>
  left_join(TABLE_PASSAGE_DF, by = c("code_insee24")) 

#EXPORT READY DATA -----
CRIMES_ARR <- CRIMES_DF |> 
  group_by(arr24, year, crimes = var_crimes) |>
  summarise(crime_count = sum(crime_count), .groups = "drop") |> 
  pivot_wider(names_from = crimes, values_from = crime_count)

CRIMES_BV22 <- CRIMES_DF |> 
  group_by(bv2022, year, crimes = var_crimes) |>
  summarise(crime_count = mean(crime_count), .groups = "drop") |> 
  pivot_wider(names_from = crimes, values_from = crime_count)

#EXPORTS ----
write_feather(as.data.frame(CRIMES_ARR), "data/interim/arrondissements/criminality.feather", compression = "zstd")
write_feather(as.data.frame(CRIMES_BV22), "data/interim/bv2022/criminality.feather", compression = "zstd")

File: src/07-B-criminality.py
import pandas as pd
import numpy as np

# Read  arrondissement‑to‑commune mapping
ARR2COM = pd.read_feather("results_building/arr2com.feather")

# Dictionary of crime codes
DICO_CRIMES = pd.DataFrame({
    "var_crimes": [
        "coups", "trafic", "vols_av", "vols_sv", "cambriolages"
    ],
    "crimes_desc": [
        "Coups et blessures volontaires",
        "Trafic de stupéfiants",
        "Vols violents sans arme",
        "Vols sans violence contre des personnes",
        "Cambriolages de logement"
    ]
})

# Build the TABLE_PASSAGE_DF analogue:
TABLE_PASSAGE_DF = (
    pd.read_feather("results_building/t_passage.feather")[["code_insee24", "arr24", "bv2022"]]
    .drop_duplicates()
    .groupby("code_insee24", as_index=False)
    .first()
)

# pre‑filter the raw crimes data
CRIMES_DF = pd.read_parquet("data/external/miom-crimes/crimes23_geo24.parquet")
CRIMES_DF = CRIMES_DF[CRIMES_DF["annee"] > 16]

# Select & clean up
CRIMES_DF = (
    CRIMES_DF[["CODGEO_2024", "annee", "faits", "classe", "unité.de.compte"]]
    .rename(columns={
        "CODGEO_2024": "code_insee24",
        "annee": "year",
        "faits": "crime_count",
        "classe": "crimes_desc",
        "unité.de.compte": "units"
    })
)
CRIMES_DF["year"] = CRIMES_DF["year"] + 2000
CRIMES_DF["crime_count"] = CRIMES_DF["crime_count"].fillna(0)
CRIMES_DF = CRIMES_DF[CRIMES_DF["crimes_desc"].isin(DICO_CRIMES["crimes_desc"])]

# Replace arr codes by commune codes
CRIMES_DF = CRIMES_DF.merge(
    ARR2COM,
    how="left",
    left_on="code_insee24",
    right_on="code_arr"
)
CRIMES_DF["code_insee24"] = np.where(
    CRIMES_DF["code_com"].notna(),
    CRIMES_DF["code_com"],
    CRIMES_DF["code_insee24"]
)

# Bring in the var_crimes code and the BV/bv2022 lookup
CRIMES_DF = CRIMES_DF.merge(DICO_CRIMES, on="crimes_desc", how="left")
CRIMES_DF = CRIMES_DF.merge(TABLE_PASSAGE_DF, on="code_insee24", how="left")

# === EXPORT DATA ===

# By arrondissement (sum of counts)
CRIMES_ARR = (
    CRIMES_DF
    .groupby(["arr24", "year", "var_crimes"], as_index=False)["crime_count"]
    .sum()
    .pivot(index=["arr24", "year"], columns="var_crimes", values="crime_count")
    .reset_index()
)

# By BV2022 (mean of counts)
CRIMES_BV22 = (
    CRIMES_DF
    .groupby(["bv2022", "year", "var_crimes"], as_index=False)["crime_count"]
    .mean()
    .pivot(index=["bv2022", "year"], columns="var_crimes", values="crime_count")
    .reset_index()
)

# Export
CRIMES_ARR.to_feather("data/interim/arrondissements/criminality.feather", compression="zstd")
CRIMES_BV22.to_feather("data/interim/bv2022/criminality.feather", compression="zstd")
